### ISSUE TITLE
Roadmap for the Project, and new Maintainer for the Plugin

### ISSUE THREAD

**Issue by joshuataylor** [KronicDeth/intellij-elixir#3598]
## Introduction

Hi All,

The IntelliJ Elixir plugin was started a decade ago by @KronicDeth, and it is a fantastic plugin with 1.8k stars on Github, 439k downloads through the JetBrains Plugin Marketplace.

The contributors have invested tremendous time, energy, and effort into creating an open-source plugin, written in Java/Kotlin, for a community primarily working in Elixir.

Unfortunately, life events happens, and opensource work is usually the last thing on the list of things that the maintainer will want to work on - I know this from personal experience where life gets really busy.

I contacted @KronicDeth , and am pleased to announce I have been added as a co-maintainer for this repository, and am excited to help contribute to the project's ongoing development.

I have been using Elixir for many years, and this plugin to write all of it :-).

## What this means

Honestly, not much will change for users of the plugin, other than more frequent updates.

I just really wanted to note that there will be no change to the license, Code of Conduct, etc.

1. All sponsorships, donations, etc will go to @KronicDeth as usual.
2. License is still `Apache License`
3. Hopefully more frequent code updates/reviews, which will foster more collaboration - not all issues are Java/Kotlin - it's also figuring out Erlang/Elixir issues :).
4. Add GitHub Discussions, which will help with questions.

## Roadmap

I would like to propose a roadmap going forward, ordered by importance.

### Phase 1: 2024 support.
1. Update to support 2024.xx supported IDEs, see [#3569](https://github.com/KronicDeth/intellij-elixir/pull/3569)
2. Review any open Pull Requests

### Phase 2: Update build environment
Ensure we've got a "modern" build/test/CI environment, which will help going forward.

This will hopefully make it easier and more transparent for new and existing contributors.

1. Update to the latest supported JetBrains JBR SDK and build against that.
2. Use [intellij-platform-gradle-plugin](https://github.com/JetBrains/intellij-platform-gradle-plugin) for building, testing, verifying, and publishing.
3. Update contributing and other documentation for the latest IntelliJ IDEA changes, tips for contributing, etc - if/as needed.

### Phase 3: Fix issues with new Elixir/Java Versions
As there have been a huge amount of changes made to both IntelliJ, Erlang and Elixir, we need to ensure we can accomodate these changes, whilst still maintaining backwards compatibility.

[JetBrains Changes](https://plugins.jetbrains.com/docs/intellij/api-changes-list.html#bundled-library-updates) 

To do this, a list will be created from the [Issues](https://github.com/KronicDeth/intellij-elixir/issues/) for bugs that occur due to IntelliJ, Erlang or Elixir version changes, example [#3362](https://github.com/KronicDeth/intellij-elixir/issues/3362).

Community involvement at this phase would be amazing, as there will be a slew of Java/Kotlin and Erlang/Elixir issues to diagnose and fix! Ever wanted to know how an IntelliJ Plugin is made? This would be a great chance to learn and contribute.

### Phase 4: Clear backlog
After Phase 3, if we're still sane, it'll be a good idea to clear out the backlog.

I would also like to implement a better way to report logs without having to make an issue through GitHub.

[JetBrains supports this](https://www.plugin-dev.com/intellij/general/error-reporting/), but we'll need to have a privacy-focused way to treat this, but this would be completely option, the the user must also click that button to report.

This would also allow us to track how severe a certain issue is, and prioritise it.

However I really want something privacy focused, and ensure it does not leak any of the users code they don't want to send. It might end up being a Sentry, or their self-hosted option if all else fails.

If you know a good privacy-focused error handling service that supports Java/Kotlin, reach out.

### JetBrains APIs

JetBrains APIs move quite fast, especially with major releases (eg 2023.xx -> 2024.xx). [2024 Incompatible Changes](https://plugins.jetbrains.com/docs/intellij/api-changes-list.html), for example shows how many changes have been made for 2024.

One on hand, yes this is good, as it allows them to add new features and provide a good experience.

On the other hand, this is tedious for plugin developers when things break, as sometimes there are major problems.

Going forward, testing using the EAP versions will be done on CI to ensure that everything works as expected.

## Feedback / Suggestions / Ideas

Please add any feedback, suggestions, ideas to this thread - the idea is to be as transparent and open as possible.

**Comment by ghenry** [comment]
Can JetBrains not help sponsor this? Or the Erlang Foundation?

How can we help?

**Comment by joshuataylor** [comment]
To be completely honest, my current (and future, at this stage) financial situation and personal life allows me the privilege to work on this plugin and other opensource work.

Once we fix the 2024.x issue and start creating a roadmap of issues, a great help by the community would be:

1. Working on and fixing issues caused by newer versions of Java/Kotlin/Erlang/Elixir
2. Reporting/triaging issues

From a sponsorship point of view, I'd rather leave that out for now - perhaps we can reach out for a grant for resources such as CI/CD hosting on a cheap provider like Hetzner etc to improve build times.

I'm also not an expert in Java/Kotlin, and would love any feedback for new PRs from the community as well.

**Comment by iampeter** [comment]
Again, thank you @joshuataylor for getting involved and for pushing the first updates. As for the roadmap, is broader support for Heex on the horizon?

**Comment by pdxmike** [comment]
I'm still getting in to Elixir, but have already really enjoyed this plugin.  Thanks for pushing this forward @joshuataylor, and thanks for all the long-term work on this @KronicDeth 👍

**Comment by gaggle** [comment]
Just fabulous news. Really echoing the ideas energy of finding ways for us to support the efforts here. I don't know Java/Kotlin but would love to find ways of supporting, if I'm not alone on that I want to gently suggest to you @joshuataylor that there might be worthwhile investments into little more than making some semi-regular "office hour" threads in GitHub or Elixir Forum or something to cover a specific detail of the plugin. 

I mean: If you did nothing else but tell the rest of us about the plugin's architecture, it's possible I and others could learn just enough to start contributing with code? Some architecture sketches, streamlining "how to get started", and other kind of almost banal grassroots-enabling things. E.g. a thread somewhere where you could bring up how to get started, and then we all just kinda try it out together and get to ask all our stupid questions?

I'm not suggesting this to take you or anyone's energy away from the code-work that is needed and is probably what most of us assume when you post about becoming a co-contributor, but I think this other angle is also worth considering.

**Comment by joshuataylor** [comment]
Yeah, I'm super keen to make it as easy as possible for people to contribute, and with the recent improvements from JetBrains around developer experience should make this a lot easier. We'e probably not Java/Kotlin developers, but Elixir developers :-).

I believe adding discussions to this repo would be good for questions around contributing, plugin questions etc, and I would love to do an office hours, let's get that started.

I honestly believe adding a feature, fixing a bug should be a really straight forward process, you can use the IntelliJ Community Edition for this. Make it as easy as possible.

I'm also still learning about the plugin architecture, but will definitely create a broad overview of this, and links back to the JetBrains documentation about that feature.

I also want to extend out the CONTRIBUTING.MD with all of this as well!

**Comment by yordis** [comment]
Second to that,

At some point, I tried my best to contribute https://github.com/KronicDeth/intellij-elixir/pull/2578. Ultimately, I had to blindly add code that I couldn't run nor understand how, primarily because of getting the devtools working and having a process around it.

So, from first-hand experience, anything you can do to document and unblock folks who are capable of figuring out Java but do not know enough about the Java ecosystem or how to get to a point where we can put a breakpoint, I think it will pay off.

**Comment by marcdel** [comment]
This is a roadmap-ish question: I'm trying to understand which things are hard/impossible with the plugin API and parser/lexer vs. just haven't been prioritized. Does anyone have a feel for whether things like the extract/inline refactorings are possible here?

**Comment by gaggle** [comment]
Hi @joshuataylor,

We're crawling into 5 months since your post and I wonder how you're feeling. I worry you get burnt out, basically :). I have email notifications on for new issues, and every time I see the automated issues tick in I think to myself I'd get burnt out if it was me. I'd need to put my foot down and automate all issues into an easily hidden triage-corner, and shine a laser-focus on the essential work that is actually going on. But I hope you're coping!

I could offer to set up such issue-automation, and wipe out Milestones and other cruft in the project, if you feel there's value in a bit of "bull in a china store" energy.

I don't mean to intrude though and I hope you're just loving the project exactly the way it is, in which case I simply wish you all the best energy possible. In any event my thanks for your ongoing support of this plugin.
